### ISSUE TITLE
UMI barcodes error

### ISSUE THREAD

**Issue by jaya-share** [uubram/RTCR#8]
Hi, I found a strange error and was wondering if I am providing the barcode input correctly as
The Checkout function gives the same output irrespective of different slave barcodes
Pool9-M1ss701_513-beta	TAAGGCGAcagtggtatcaacgcagagtNNNNtNNNNtNNNNtct	TCGACTAGacacsttkttcaggtcctc
Pool9-M1ss702_513-beta	CGTACTAGcagtggtatcaacgcagagtNNNNtNNNNtNNNNtct	TCGACTAGacacsttkttcaggtcctc

The same output files are produced with the following barcodes as well irrespective the slave are different..
Pool9-M1ss701_515-beta	TAAGGCGAcagtggtatcaacgcagagtNNNNtNNNNtNNNNtct	TTCTAGCTacacsttkttcaggtcctc
Pool9-M1ss702_515-beta	CGTACTAGcagtggtatcaacgcagagtNNNNtNNNNtNNNNtct	TTCTAGCTacacsttkttcaggtcctc

Your feedback will be quite useful..please suggest the right way.

**Comment by uubram** [comment]
Slave barcodes are essentially only used to retrieve the remainder of the UMI. Since your slave barcodes have no UMI, the output is unaffected. In your message you used space as delimiter, I presume you are using tab in the actual barcode files.

**Comment by jaya-share** [comment]
hi, yes, I have used tab delimiter in the barcode file..Can you suggest what is the best way to process such samples where master is same but slave differs, as I feel the output is not considering slave at all and all files generated are same.
thanks

**Comment by jaya-share** [comment]
This barcode file works and generate output with MIGEC but I want to try RTCR ..thus, I am keen to make it work. Any suggestion on how it can be done is really appreciated

**Comment by uubram** [comment]
Thank you for noticing the issue with adapters that have the same master sequence, but different slave sequences. I have fixed this issue in RTCR and updated the repository (it is now version 0.4.5).

The adapter sequences you showed contained IUPAC codes. I have updated RTCR so that it will handle that type of adapter sequence. Note, the seed sequences should only contain unambiguous DNA bases (i.e. A, C, T, and G).

I hope this solved your issue.

**Comment by jaya-share** [comment]
Hi, Thanks for looking into and the update..However, I am not able to fetch the latest version,, the version getting installed is (Finished processing dependencies for rtcr==0.4.4). even, git pull is not updating to 0.4.5..can you please suggest

**Comment by uubram** [comment]
My apologies, I neglected to update the version number. Hopefully it works now.

**Comment by jaya-share** [comment]
Thanks again!! the version got updated but I still I get the same output file with different slave barcodes..if you can provide me an email id I can share sample file

**Comment by uubram** [comment]
Is it possible to create a small example that highlights the issue? You can attach a file in a comment in this thread.

**Comment by jaya-share** [comment]
Hi there, Sorry I tried but can't upload as the raw file size exceeds 10MB..I don't really know how to trim the input file even assembled file is nearly 50MB

**Comment by uubram** [comment]
You can upload it to the following link:
https://www.dropbox.com/request/lBCj3VVSGWJaYGMm7APT

**Comment by jaya-share** [comment]
Hi,
thanks a lot for looking into this, I have put the raw files and the
assembled one but since the assembled one is quite small if you can check
for the raw file for 513.txt and 515.txt as I get the same output.

Thanks
Sincerely,
Jaya

**Comment by uubram** [comment]
I think I understand the issue you are running into. In short, to solve it you should combine your 513.txt and 515.txt into a single file and then analyze the fastq files only once. Below I will explain in more detail what is going on.

Apparently you are running RTCR twice on the same fastq files but supplying different adapters both times (the 513.txt and 515.txt you mentioned). The difference between the adapters in both files is that only the slave sequences differ between runs. The reason you are getting the same output is because the "checkout" algorithm tries to assign the best matching adapter to each record in the fastq files, irrespective of having a match to the slave sequence in the adapter. So let's say RTCR is processing a fastq record that should be assigned to an adapter in 515.txt, but you supplied only 513.txt, it will assign it to the best matching adapter in 513.txt (which it can because the master sequences are the same between both adapter files). The second time you run it and supply 515.txt, the same thing will happen but the other way around, hence the identical output.

I hope this solved your issue.

**Comment by jaya-share** [comment]
Hi there,
I really appreciate your time and effort to solve my query. I agree that I
separated the files when I was not able to get any output just to check.
But, now after your reply when I combine the files and take a look just
considering assembled.fastq file I observe another pattern.
If I concatenate 513 followed by 515 in as barcode file I get fasta
populated for 513 where as if I change the order(515 then 513) then 515
files are generated.

I am attaching the screen short to have a look.

Thanks
Sincerely
Jaya

**Comment by uubram** [comment]
I don't see the attached screenshot, but when I run it on your files, I get files for both 515 and 513.

**Comment by jaya-share** [comment]
Please find the files attached..if you get the files of reasonable size then I must be running the command wrong. the command I ran
rtcr Checkout -p Dpool.assembled.fastq -rc -b D_beta2.txt
or can you share the command you ran please

<img width="982" alt="00_513 then 515 " src="https://user-images.githubusercontent.com/39492778/79272242-12e03100-7e99-11ea-82b8-110b6ffe552f.png">

<img width="1219" alt="01 515 then 513" src="https://user-images.githubusercontent.com/39492778/79272284-2390a700-7e99-11ea-89d3-54fd1e86b9d0.png">

**Comment by uubram** [comment]
Thank you for the screenshots. I was able to reproduce your issue where the order of the adapters influences the output. This came about due to the following behaviors of the "checkout" algorithm:
1) it did not search for slave sequences in the reverse orientation of the master sequence
2) it assigned fastq records to an adapter even if no slave sequence was found

I decided to adjust the algorithm so that it will search for slave sequences in the reverse orientation of the master sequence (at least when running with the "-rc" option). Second, an adapter will no longer get a fastq record assigned if no slave sequence was found (and a slave sequence was defined for the adapter). These changes have been introduced into the latest version (0.4.6) which should be online now.

I hope this solves the issue.
